lib/bolt/module_installer/specs/git_spec.rb

```diff
@@ -0,0 +1,173 @@
+        end
+
+        # Returns true if the specification is satisfied by the module.
+        #
+        def satisfied_by?(mod)
+          @type == mod.type && @git == mod.git
+        end
+
+        # Returns a hash matching the module spec in bolt-project.yaml
+        #
+        def to_hash
+          {
+            'git' => @git,
+            'ref' => @ref
+          }
+        end
+
+        # Returns a PuppetfileResolver::Model::GitModule object for resolving.
+        #
+        def to_resolver_module
+          require 'puppetfile-resolver'
+
+          PuppetfileResolver::Puppetfile::GitModule.new(name).tap do |mod|
+            mod.remote = @git
+            mod.ref    = sha
+          end
+        end
+
+        # Resolves the module's title from the module metadata. This is lazily
+        # resolved since Bolt does not always need to know a Git module's name.
+        #
+        def name
+          @name ||= begin
+            url      = "https://raw.githubusercontent.com/#{@repo}/#{sha}/metadata.json"
+            response = make_request(:Get, url)
+
+            case response
+            when Net::HTTPOK
+              body = JSON.parse(response.body)
+
+              unless body.key?('name')
+                raise Bolt::Error.new(
+                  "Missing name in metadata.json at #{git}. This is not a valid module.",
+                  "bolt/missing-module-name-error"
+                )
+              end
+
+              parse_name(body['name'])
+            else
+              raise Bolt::Error.new(
+                "Missing metadata.json at #{git}. This is not a valid module.",
+                "bolt/missing-module-metadata-error"
+              )
+            end
+          end
+        end
+
+        # Resolves the SHA for the specified ref. This is lazily resolved since
+        # Bolt does not always need to know a Git module's SHA.
+        #
+        def sha
+          @sha ||= begin
+            url      = "https://api.github.com/repos/#{@repo}/commits/#{ref}"
+            headers  = ENV['GITHUB_TOKEN'] ? { "Authorization" => "token #{ENV['GITHUB_TOKEN']}" } : {}
+            response = make_request(:Get, url, headers)
+
+            case response
+            when Net::HTTPOK
+              body = JSON.parse(response.body)
+              body['sha']
+            when Net::HTTPForbidden
+              message = "GitHub API rate limit exceeded, unable to resolve git modules. "
+
+              unless ENV['GITHUB_TOKEN']
+                message += "To increase your rate limit, set the GITHUB_TOKEN environment "\
+                          "variable with a GitHub personal access token."
+              end
+
+              raise Bolt::Error.new(message, 'bolt/github-api-rate-limit-error')
+            when Net::HTTPNotFound
+              raise Bolt::Error.new(
+                "#{git} is not a Git repository.",
+                "bolt/missing-git-repository-error"
+              )
+            else
```

Comment:
This can also be a 401 `Net::HTTPUnauthorized ` if the `GITHUB_TOKEN` is invalid